File: sdk/typescript/packages/channel-telegram/src/formatting.ts
export function markdownToTelegramHtml(text: string): string {
  if (!text) {
    return "";
  }

  const codeBlocks: string[] = [];
  const inlineCodes: string[] = [];
  const linkHrefs: string[] = [];

  let output = text.replace(/```[\w]*\n?([\s\S]*?)```/g, (_match, code: string) => {
    codeBlocks.push(code);
    return `\u0000CB${codeBlocks.length - 1}\u0000`;
  });

  output = output.replace(/`([^`]+)`/g, (_match, code: string) => {
    inlineCodes.push(code);
    return `\u0000IC${inlineCodes.length - 1}\u0000`;
  });

  output = output.replace(/^#{1,6}\s+(.+)$/gm, "$1");
  output = output.replace(/^>\s*(.*)$/gm, "$1");
  output = output.replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;");
  output = output.replace(/\[([^\]]+)\]\(([^)]+)\)/g, (_match, linkText: string, href: string) => {
    linkHrefs.push(href);
    return `<a href="\u0000LK${linkHrefs.length - 1}\u0000">${linkText}</a>`;
  });
  output = output.replace(/\*\*(.+?)\*\*/g, "<b>$1</b>");
  output = output.replace(/__(.+?)__/g, "<b>$1</b>");
  output = output.replace(/(?<![a-zA-Z0-9])_([^_]+)_(?![a-zA-Z0-9])/g, "<i>$1</i>");
  output = output.replace(/~~(.+?)~~/g, "<s>$1</s>");
  output = output.replace(/^[-*]\s+/gm, "• ");

  for (let index = 0; index < inlineCodes.length; index += 1) {
    const escaped = escapeHtml(inlineCodes[index] ?? "");
    output = output.replace(`\u0000IC${index}\u0000`, `<code>${escaped}</code>`);
  }

  for (let index = 0; index < codeBlocks.length; index += 1) {
    const escaped = escapeHtml(codeBlocks[index] ?? "");
    output = output.replace(`\u0000CB${index}\u0000`, `<pre><code>${escaped}</code></pre>`);
  }

  for (let index = 0; index < linkHrefs.length; index += 1) {
    output = output.replace(`\u0000LK${index}\u0000`, linkHrefs[index] ?? "");
  }

  return output;
}

export function splitTelegramMessage(content: string, maxLength = 4000): string[] {
  if (content.length <= maxLength) {
    return [content];
  }

  const chunks: string[] = [];
  let remaining = content;

  while (remaining.length > 0) {
    if (remaining.length <= maxLength) {
      chunks.push(remaining);
      break;
    }

    const cut = remaining.slice(0, maxLength);
    let position = cut.lastIndexOf("\n\n");
    if (position === -1) {
      position = cut.lastIndexOf("\n");
    }
    if (position === -1) {
      position = cut.lastIndexOf(" ");
    }
    if (position === -1) {
      position = maxLength;
    }

    chunks.push(remaining.slice(0, position).trimEnd());
    remaining = remaining.slice(position).trimStart();
  }

  return chunks;
}

function escapeHtml(value: string): string {
  return value.replaceAll("&", "&amp;").replaceAll("<", "&lt;").replaceAll(">", "&gt;");
}

File: sdk/python/examples/telegram/dotcraft_telegram/formatting.py
"""Telegram message formatting utilities."""

from __future__ import annotations

import re


def markdown_to_telegram_html(text: str) -> str:
    """
    Convert Markdown to Telegram-safe HTML.

    Telegram's HTML parse mode supports: <b>, <i>, <s>, <u>, <code>,
    <pre>, <a href="...">, and basic entity escaping (&amp; &lt; &gt;).

    Strategy:
    1. Extract and protect code blocks / inline code (preserve from escaping).
    2. Convert structural elements (headers, blockquotes, lists).
    3. HTML-escape the remaining text.
    4. Apply inline formatting (bold, italic, strikethrough, links).
    5. Restore protected code blocks.
    """
    if not text:
        return ""

    # --- Step 1: protect code blocks ---
    code_blocks: list[str] = []

    def save_code_block(m: re.Match) -> str:
        code_blocks.append(m.group(1))
        return f"\x00CB{len(code_blocks) - 1}\x00"

    text = re.sub(r"```[\w]*\n?([\s\S]*?)```", save_code_block, text)

    # --- Step 2: protect inline code ---
    inline_codes: list[str] = []

    def save_inline_code(m: re.Match) -> str:
        inline_codes.append(m.group(1))
        return f"\x00IC{len(inline_codes) - 1}\x00"

    text = re.sub(r"`([^`]+)`", save_inline_code, text)

    # --- Step 3: strip header markers (keep text) ---
    text = re.sub(r"^#{1,6}\s+(.+)$", r"\1", text, flags=re.MULTILINE)

    # --- Step 4: strip blockquote markers ---
    text = re.sub(r"^>\s*(.*)$", r"\1", text, flags=re.MULTILINE)

    # --- Step 5: HTML-escape ---
    text = text.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")

    # --- Step 6: links [text](url) ---
    text = re.sub(r"\[([^\]]+)\]\(([^)]+)\)", r'<a href="\2">\1</a>', text)

    # --- Step 7: bold **text** or __text__ ---
    text = re.sub(r"\*\*(.+?)\*\*", r"<b>\1</b>", text)
    text = re.sub(r"__(.+?)__", r"<b>\1</b>", text)

    # --- Step 8: italic _text_ (avoid matching inside snake_case words) ---
    text = re.sub(r"(?<![a-zA-Z0-9])_([^_]+)_(?![a-zA-Z0-9])", r"<i>\1</i>", text)

    # --- Step 9: strikethrough ~~text~~ ---
    text = re.sub(r"~~(.+?)~~", r"<s>\1</s>", text)

    # --- Step 10: bullet lists ---
    text = re.sub(r"^[-*]\s+", "• ", text, flags=re.MULTILINE)

    # --- Step 11: restore inline code ---
    for i, code in enumerate(inline_codes):
        escaped = code.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
        text = text.replace(f"\x00IC{i}\x00", f"<code>{escaped}</code>")

    # --- Step 12: restore code blocks ---
    for i, code in enumerate(code_blocks):
        escaped = code.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
        text = text.replace(f"\x00CB{i}\x00", f"<pre><code>{escaped}</code></pre>")

    return text


def split_message(content: str, max_len: int = 4000) -> list[str]:
    """
    Split a long message into chunks that fit within Telegram's message limit.

    Prefers splitting at paragraph breaks, then line breaks, then word
    boundaries, and falls back to hard-cutting at max_len characters.
    """
    if len(content) <= max_len:
        return [content]

    chunks: list[str] = []
    while content:
        if len(content) <= max_len:
            chunks.append(content)
            break
        cut = content[:max_len]
        # Prefer paragraph break
        pos = cut.rfind("\n\n")
        if pos == -1:
            pos = cut.rfind("\n")
        if pos == -1:
            pos = cut.rfind(" ")
        if pos == -1:
            pos = max_len
        chunks.append(content[:pos].rstrip())
        content = content[pos:].lstrip()

    return chunks
